subsys/bluetooth/host/l2cap.c

```diff
@@ -2973,6 +2975,41 @@ int bt_l2cap_ecred_chan_reconfigure(struct bt_l2cap_chan **chans, uint16_t mtu)
 		if (mtu < BT_L2CAP_LE_CHAN(chans[i])->rx.mtu) {
 			return -EINVAL;
 		}
+
+#if defined(CONFIG_BT_L2CAP_SEG_RECV)
+		/**
+		 * If more than one channel is being configured, the MPS field
+		 * shall be greater than or equal to the current MPS size of
+		 * each of these channels. If only one channel is being
+		 * configured, the MPS field may be less than the current MPS
+		 * of that channel.
+		 */
+		if (BT_L2CAP_LE_CHAN(chans[i])->chan.ops->seg_recv) {
+			seg_direct = true;
+		}
+#endif
+
+		if (i == 0) {
+			mps = BT_L2CAP_LE_CHAN(chans[i])->rx.mps;
+		}
+
+		if (i > 0) {
+			/* For channels using segment direct API we can only
+			 * decrease MPS and this is not allowed for multiple
+			 * channels. So if any channel is using this MPS cannot
+			 * change.
+			 *
+			 * TODO it should be possible to increase if previously
+			 * decreased but this would require extra tracking
+			 */
+			if (seg_direct) {
+				if (mps != BT_L2CAP_LE_CHAN(chans[i])->rx.mps) {
+					return -EINVAL;
+				}
+			}
+
+			mps = MAX(mps, BT_L2CAP_LE_CHAN(chans[i])->rx.mps);
```

Comment:
The MPS should not be the maximum. The MPS in the reconfigure request is the new MPS for all the listed channels.

The right thing to do here is to error out if the wanted MPSs are not all equal.

Note, the MPS of non-seg_direct channels is always `MIN(BT_L2CAP_LE_CHAN(chans[i])->rx.mps, BT_L2CAP_RX_MTU)` and does not ever change.

--------

I think we should be able to add this assert, if it helps:

`!BT_L2CAP_LE_CHAN(chans[i])->chan.ops->recv || BT_L2CAP_LE_CHAN(chans[i])->rx.mps == MIN(BT_L2CAP_LE_CHAN(chans[i])->rx.mps, BT_L2CAP_RX_MTU)`

Here, the MPS should ideally be just equal to `BT_L2CAP_RX_MTU`. But for some reason we complicate this in Zephyr. See the following code comment, that's not citing any source for its claim.

https://github.com/zephyrproject-rtos/zephyr/blob/e7db0f8aff81436ff07f557a5ca8e2daa1789188/subsys/bluetooth/host/l2cap.c#L1226-L1230

Reply:
> and does not ever change.

Actually, that part is wrong, since we do the complicated calculation. It would be right if it just always stayed at `BT_L2CAP_RX_MTU`. Fixed.

Reply:
normally all channels have same MPS and exception are only channels that use segment direct API,    currently on reconfigure such channels will be reset to BT_L2CAP_RX_MTU which might be bigger than what was set by applications.  This MAX() is to ensure that we don't unnecessary clip MPS, but yeah maybe I could just assert if those are not same since  channels with seg direct API are already handled above

BTW I'm not sure why MPS is configurable in segmented direct API channels, instead of simply using same MPS as all other channels.




regarding clipping MPS to MTU...
not sure if there are other IUTs that have this problem, but at least PTS gets triggered by this,  ie if MTU<MPS it may send SDU>MTU in some tests (probably due to hardcoded values),    this means extra work reporting that to PTS folks, testing their patches and broken tests until PTS with fix included is released

Reply:
> BTW I'm not sure why MPS is configurable in segmented direct API channels, instead of simply using same MPS as all other channels.

The reason I made it configurable is that setting the MPS gives application services some control over what one credit is worth. Say the application is able to receive a maximum of 128 bytes of data at the moment. If the MPS is set to 64, it can send 2 credits. If the MPS is set globally to e.g. 256, it can't safely send even one credit.

A global setting is not very flexible when multiple protocols use L2CAP. You may want one MPS for ATT, and another for other uses.

But I am the first to say that I'm not sure how the designers of L2CAP credits envisioned this will be used. It's a awkward system in conjunction with SDUs, which 1. "steals" two bytes of the MPS in the first segment for the SDU length header, and 2. you don't really know how many credits you need to complete an SDU.